Refactor/make it go

RobjS opened the pull request:
I think this is ready for a look now. I've just duplicated the existing functionality, with one important exception: the command used to run wpc.

Previously:

```
wpc_init my-project --multisite
```

Now:

```
wpc -project=my-project -multisite
```

As far as I can tell, the command will always be the same as the package directory (hence `wpc`). And I couldn't find a way to get Go to accept arguments/flags in the format we were previously using (see https://stackoverflow.com/questions/25248534/golang-flag-parsing-after-an-argument-on-command-line).  But if there's a solution to either of those issues, happy to use it.

mallorydxw commented on README.md:
````diff
@@ -4,15 +4,18 @@
 ```
-git clone https://github.com/dxw/wpc
-sudo cp wpc/bin/* /usr/local/bin/
+go get https://github.com/dxw/wpc
````
That should be `go get github.com/dxw/wpc`

mallorydxw commented on main.go:
```diff
@@ -0,0 +1,110 @@
+
+func main() {
+    /* Go won't allow the same arg/flag construction we had before, so all flags instead */
+    project := flag.String("project", "", "Project name")
```
You can fetch the non-flag arguments with this function: https://godoc.org/flag#Args

RobjS replied:
Done.

mallorydxw commented on main.go:
```diff
@@ -0,0 +1,110 @@
+    "os"
+    "io/ioutil"
+    "bytes"
+    "flag"
```
This is supposed to be a drop-in replacement for the `flag` package: https://godoc.org/github.com/spf13/pflag

So just replace

```
"flag"
```

with

```
flag "github.com/spf13/pflag"
```

and we should be able to use `--multisite` instead of `-multisite`.

(I haven't used this package before. There may be better ones)

RobjS replied:
Done.

mallorydxw commented on main.go:
```diff
@@ -0,0 +1,110 @@
+    multisite := flag.Bool("multisite", false, "Make it a multisite")
+    flag.Parse()
+    if *project == "" {
+        println("You must set a value for the project argument")
```
Use `fmt.Println` instead of `println`.

https://golang.org/ref/spec#Bootstrapping

RobjS replied:
Done.

mallorydxw commented on main.go:
```diff
@@ -0,0 +1,110 @@
+    makedir("setup/content")
+
+    createDockerCompose(*project)
+    createBinWp()
```
Instead of creating lots of tiny functions, it might make sense to just do `creating("bin/wp", wpContents())` here?

RobjS replied:
Done.

mallorydxw commented on main.go:
```diff
@@ -0,0 +1,110 @@
+func creating(file string, data[]byte) {
+    creatingNoexec(file, data)
+    os.Chmod(file, 0755)
+}
```
We could probably merge `creating` and `creatingNoexec` into one function which accepts a `os.FileMode`.

Then create couple of constants like `const NORMALFILE = 0644` `const NOEXEC = 0755`.

Then instead of `creating(...)` and `creatingNoexec(...)` we could use `creating(..., NORMALFILE)` and `creating(..., NOEXEC)`.

RobjS replied:
Done.

mallorydxw commented on main.go:
```diff
@@ -0,0 +1,110 @@
+
+    makedir("bin")
+    makedir("setup")
+    makedir("setup/content")
```
I think it would make sense to move the calls to `makedir()` into `creating()`. The function to fetch the directory name from a path is: https://godoc.org/path#Dir

RobjS replied:
That would make sense, but one of the directories (`setup/content`) doesn't actually receive any files, so never gets referenced by `creating()`. So instead, I've tried to simplify this a bit by just setting up an array of required directories, then just calling `makeDirs()` once with that array. I've also used `MkdirAll()` within that function so that subdirectories don't have to be created separately.

mallorydxw commented on main.go:
```diff
@@ -0,0 +1,110 @@
+
+func makedir(directory string) {
+    println("-> Creating", directory)
+    os.Mkdir(directory, 0755)
```
If we call `makedir()` from `creating()` then it would make sense to use `os.MkdirAll()` here. So we don't need to worry about the order things are created in.

mallorydxw commented on setupContents.go:
```diff
@@ -0,0 +1,80 @@
+
+if wp theme is-installed $theme
+then
+  !!!THEMEENABLE!!!
```
Check out the examples here: https://godoc.org/text/template#Template

RobjS replied:
Still working on this one.

mallorydxw commented on setupContents.go:
```diff
@@ -0,0 +1,80 @@
+package main
+
+func externalContents() []byte {
```
Why use a function here? These could be constants easily enough. Also, `string` is usually easier to work with than `[]byte`. I typically only convert my `string`s to `[]byte`s when writing them to files.

RobjS replied:
Done.

mallorydxw commented on bin/wpc_init:
```diff
@@ -1,277 +0,0 @@
-#!/bin/sh
```
Bye shell script. 👋

RobjS commented:
@tomdxw Ok, I think this is ready for another look now.

RobjS commented:
Ah yes, good spot. Should be fixed now.

mallorydxw commented:
Sorry I keep coming back to you with "one last thing".

When using `--multisite`, the file `config/server.php` doesn't get generated.

RobjS commented:
My bad, I should have picked that up. Added now. I also removed the `$base` variable set in that file, as I couldn't see that it was doing anything, but let me know if it has a purpose I've missed.
